substrate/pkg/controller/substrate/cluster/config.go

```diff
@@ -0,0 +1,259 @@
+	defaultStaticConfig.ClusterConfiguration.ImageRepository = imageRepository
+	defaultStaticConfig.Etcd.Local = &kubeadm.LocalEtcd{
+		ServerCertSANs: []string{"localhost", "127.0.0.1"},
+		PeerCertSANs:   []string{"localhost", "127.0.0.1"},
+		DataDir:        "/var/lib/etcd",
+		ExtraArgs: map[string]string{
+			"initial-cluster":             fmt.Sprintf("%s=https://127.0.0.1:2380", substrate.Name),
+			"initial-cluster-state":       "new",
+			"name":                        substrate.Name,
+			"listen-peer-urls":            "https://127.0.0.1:2380",
+			"listen-client-urls":          "https://127.0.0.1:2379",
+			"advertise-client-urls":       "https://127.0.0.1:2379",
+			"initial-advertise-peer-urls": "https://127.0.0.1:2380",
+		},
+	}
+	// master specific config
+	masterElasticIP := aws.StringValue(substrate.Status.Cluster.Address)
+	defaultStaticConfig.LocalAPIEndpoint.AdvertiseAddress = masterElasticIP
+	defaultStaticConfig.LocalAPIEndpoint.BindPort = 443
+	defaultStaticConfig.ControlPlaneEndpoint = masterElasticIP + ":443"
+	defaultStaticConfig.APIServer.CertSANs = []string{masterElasticIP, substrate.Name,
+		"kubernetes", "kubernetes.default", "kubernetes.default.svc", "kubernetes.default.svc.cluster.local", "10.96.0.1"}
+	defaultStaticConfig.APIServer.ExtraArgs = map[string]string{
+		"advertise-address": masterElasticIP,
+		"secure-port":       "443",
+	}
+	if defaultStaticConfig.Scheduler.ExtraArgs == nil {
+		defaultStaticConfig.Scheduler.ExtraArgs = map[string]string{}
+	}
+	if defaultStaticConfig.ControllerManager.ExtraArgs == nil {
+		defaultStaticConfig.ControllerManager.ExtraArgs = map[string]string{}
+	}
+	defaultStaticConfig.NodeRegistration = kubeadm.NodeRegistrationOptions{Name: substrate.Name}
+	return defaultStaticConfig, nil
+}
+
+// DirectoryIterator represents an iterator of a specified directory
+type DirectoryIterator struct {
+	filePaths []string
+	bucket    string
+	next      struct {
+		path string
+		f    *os.File
+	}
+	err error
+}
+
+// NewDirectoryIterator builds a new DirectoryIterator
+func NewDirectoryIterator(bucket, dir string) s3manager.BatchUploadIterator {
+	var paths []string
+	filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			paths = append(paths, path)
+		}
+		return nil
+	})
+	return &DirectoryIterator{
+		filePaths: paths,
+		bucket:    bucket,
+	}
+}
+
+// Next returns whether next file exists or not
+func (d *DirectoryIterator) Next() bool {
+	if len(d.filePaths) == 0 {
+		d.next.f = nil
+		return false
+	}
+	f, err := os.Open(d.filePaths[0])
+	d.err = err
+	d.next.f = f
+	d.next.path = d.filePaths[0]
+	d.filePaths = d.filePaths[1:]
+	return true && d.Err() == nil
+}
+
+// Err returns error of DirectoryIterator
+func (d *DirectoryIterator) Err() error {
+	return d.err
+}
+
+// UploadObject uploads a file
+func (d *DirectoryIterator) UploadObject() s3manager.BatchUploadObject {
```

Comment:
Does this upload in parallel?

Reply:
>The SDK enables you to pass custom iterators to the new batched operations. For example, if we want to upload a directory, none of the default iterators do this easily. The following example shows how to implement a custom iterator that uploads a directory to S3.

Man -- instead of writing docs, why not write a directory iterator :O